pkg/reconciler/common/prune.go

```diff
@@ -21,75 +21,199 @@ import (
 	for _, ns := range nsList.Items {
 		if ignore := re.MatchString(ns.GetName()); ignore {
 			continue
 		}
-		allNameSpaces = append(allNameSpaces, ns.Name)
+		nsAnnotations := ns.GetAnnotations()
+		//skip all the namespaces if annotated with prune skip
+		if nsAnnotations[pruneSkip] != "" && nsAnnotations[pruneSkip] == "true" {
+			continue
+		}
+		pc := &pruneConfigPerNS{
+			config:               v1alpha1.Prune{},
+			relatedPipelineNames: []string{},
+			relatedTaskNames:     []string{},
+		}
+
+		if nsAnnotations[pruneResourcesPipelineNames] != "" {
+			pc.relatedPipelineNames = strings.Split(nsAnnotations[pruneResourcesPipelineNames], ",")
+		}
+		if nsAnnotations[pruneResourcesTaskNames] != "" {
+			pc.relatedTaskNames = strings.Split(nsAnnotations[pruneResourcesTaskNames], ",")
+		}
+		if nsAnnotations[pruneResources] != "" {
+			pc.config.Resources = strings.Split(nsAnnotations[pruneResources], ",")
+		} else {
+			pc.config.Resources = defaultPruneConfig.Resources
+		}
+		if nsAnnotations[pruneKeepSince] != "" {
+			keepsince, _ := strconv.Atoi(nsAnnotations[pruneKeepSince])
+			uintKeepSince := uint(keepsince)
+			pc.config.KeepSince = &uintKeepSince
+			pc.config.Keep = nil
+		} else if defaultPruneConfig.KeepSince != nil && nsAnnotations[pruneKeep] == "" {
+			pc.config.KeepSince = defaultPruneConfig.KeepSince
+		}
+		if nsAnnotations[pruneKeep] != "" {
+			keep, _ := strconv.Atoi(nsAnnotations[pruneKeep])
+			uintKeep := uint(keep)
+			pc.config.Keep = &uintKeep
+			pc.config.KeepSince = nil
+		} else if defaultPruneConfig.Keep != nil && nsAnnotations[pruneKeepSince] == "" {
+			pc.config.Keep = defaultPruneConfig.Keep
+		}
+
+		// if a different schedule then create a new cronJob
+		if nsAnnotations[pruneSchedule] != "" {
+			if nsAnnotations[pruneSchedule] != defaultPruneConfig.Schedule {
+				pc.config.Schedule = nsAnnotations[pruneSchedule]
+				uniqueSchedule[ns.Name] = pc
+				delete(commonSchedule, ns.Name)
+				continue
+			}
+		}
+
+		commonSchedule[ns.Name] = pc
 	}
-	return allNameSpaces, nil
+	prunableNs.commonScheduleNs = commonSchedule
+	prunableNs.uniqueScheduleNS = uniqueSchedule
+	return prunableNs, nil
 }
 
-func createCronJob(k kubernetes.Interface, ctx context.Context, pru v1alpha1.Prune, targetNs string, pruningNs []string, oR v1.OwnerReference, tknImage string) error {
-	pruneContainers := getPruningContainers(pru.Resources, pruningNs, *pru.Keep, tknImage)
+func (pruner *Pruner) createAllJobs(nsConfig map[string]*pruneConfigPerNS) []corev1.Container {
+	var containers []corev1.Container
+	for ns, con := range nsConfig {
+		jobContainers := pruner.createJobs(con, ns)
+		containers = append(containers, jobContainers...)
+	}
+	return containers
+}
+
+func (pruner *Pruner) createJobs(nsConfig *pruneConfigPerNS, ns string) []corev1.Container {
+	var containers []corev1.Container
+
+	cmdArgs := pruneCommand(nsConfig, ns)
+	jobName := SimpleNameGenerator.RestrictLengthWithRandomSuffix("pruner-tkn-" + ns)
```

Comment:
let us avoid the `namespace` suffix. it is redundant data and it doesn't help us with any functional benefits.

Im assuming that this was done to check a cronjob was created by this pruner logic. For that, let us add a label on the CornJob (example `tektonconfig.operator.tekton.dev/pruner: true`) 

This will be more readable. It will also help us filter our query for cronJobs from allnamespaces.

Reply:
In addition, we will create only 1 pruner CronJob in any given namespace.